release_dashboard/lib/widgets/common/continue_button.dart

```diff
@@ -0,0 +1,72 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+
+typedef DynamicFuture = Future<dynamic> Function();
+
+/// The continue button of each step.
+///
+/// [elevatedButtonKey] is assigned to [ElevatedButton]. It makes locating and
+/// pressing the button easier in tests.
+///
+/// [error] is the error string that gets displayed when there is an error in the release.
+/// procedure. The error text is in red and displayed above [ElevatedButton].
+///
+/// When [enabled] is true, [ElevatedButton] is enabled, disabled otherwise.
+///
+/// [onPressedCallback] will be executed when [ElevatedButton] is enabled and pressed.
+/// [onPressedCallback] must be asynchronous.
+///
+/// when [isLoading] is true, [CircularProgressIndicator] will be dispalyed, hidden otherwise.
+class ContinueButton extends StatelessWidget {
+  const ContinueButton({
+    Key? key,
+    required this.elevatedButtonKey,
+    this.error,
+    required this.enabled,
+    required this.onPressedCallback,
+    required this.isLoading,
+  }) : super(key: key);
+
+  final Key elevatedButtonKey;
+  final String? error;
+  final bool enabled;
+  final DynamicFuture onPressedCallback;
+  final bool isLoading;
+
+  @override
+  Widget build(BuildContext context) {
+    return Column(
+      children: [
+        if (error != null)
+          Center(
+            child: SelectableText(
+              error!,
+              style: Theme.of(context).textTheme.subtitle1!.copyWith(color: Colors.red),
+            ),
+          ),
+        Row(
+          mainAxisAlignment: MainAxisAlignment.center,
+          children: <Widget>[
+            ElevatedButton(
+              key: elevatedButtonKey,
+              onPressed: enabled == true
+                  ? () async {
+                      await onPressedCallback();
```

Comment:
This should be

```
onPressedCallback().whenComplete(() => isLoading = false);
```

In this case, the UI doesn't need to wait hang for this action to be complete, and it will update itself once the future is complete.

Reply:
Ok, is it also possible to set `isLoading` to true when there is `onPressedCallback()` is started or `onPressedCallback()` encounters an error?

Reply:
```
isLoading = true;
onPressedCallback().whenComplete(() => isLoading = false);
```

You can implement [catchError](https://api.flutter.dev/flutter/dart-async/Future/catchError.html), but IIRC `whenComplete` acts like `finally` so it is guaranteed to run last.

Reply:
I realized I can't offset the `isLoading` and `error` to `ContinueButton`, they will have to be passed by the parents. Since the parent widget is using `isLoading` to disable the inputs and dropdowns, and needs to modify `error` to display other errors not thrown by the callbacks of `ContinueButton`.

Reply:
I changed the code to use `whenComplete` and `catchError`, please take another look.

Reply:
This widget is missing the logic (it's still in the parent `CreateReleaseSubsteps`). Leaving it in the parent means we'll need to duplicate it everywhere else we use ContinueButton

Reply:
But the parents needs `isLoading` to disable the dropdowns and inputforms. And the parent widget might also updates `_error`. If I offload the logic to the child, I will have to pass two additional parameters `setError` and `setIsLoading`.

Would it be ok to leave the logic as a callback to be passed to the child? In this way we could customize the `onPressedCallback`, and make the widget more flexible to accommodate for future changes. Because every step of the release will be using the same `ContinueButton` widget. If I bring the logic of `CreateReleaseSubsteps` inside the `ContinueButton`, the other steps are not guaranteed to have the same logic as `CreateReleaseSubsteps`. For example, another step could have a `dialoguePromp` appear as `onPressedCallback`, if `Yes` is clicked on the prompt, then executes the rest of the callback, otherwise do not execute.

Reply:
Adding the additional callbacks sgtm. This is a good example of a well defined api that can be used everywhere

Reply:
By adding the additional callbacks, do you mean by passing `setError` and `setIsLoading` as additional parameters of  `ContinueButton`? This wouldn't be necessary if we keep the logic in the parent, since `setError` and `setIsLoading` are parts of the logic of `onPressedCallback` which is defined in the parent, and passed to the child.

The main concern is having the logic declared in the parent and pass as the parameter `onPressedCallback` to the child. This way, `onPressedCallback` is flexible, since the parent can define it however it wants. If we offload it to the child `ContinueButton`, we are bound to use every instance of `ContinueButton` (used by every step) according to the logic `ContinueButton` defined, which will be difficult to accommodate for different behaviours of all continue buttons of each step, for example the one mentioned above.

Reply:
The goal with any API is to make it as intuitive as possible for users to pick up. The assumption you're baking in makes the `ContinueButton`'s onPressed to be very complicated. It's easier to use if you explicitly enable `onError` and `onLoading` as optional parameters in `ContinueButton`.

In general, it's easier to have an API like
``` dart
// Better API design where its explicit
ContinueButton({
  Future<void> onPressed,
  onError?,
  onLoading?,
})
```

versus 

``` dart
/// It's easy to invoke this widget, but it's requiring everywhere else to implement the future logic.
///
/// [onPressed] is a future that should have error handles specified before being passed.
///
/// An example for [onPressed] with error handling and loading:
///
/// ```
///.   myOnPressedFuture.catchError((error, stacktrace) {
///       setError(errorToString(error, stacktrace));
///     }).whenComplete(() {
///       setIsLoading(false);
///       if (_error == null) {
///         myNextFuture();
///       }
///     });
/// ```
///
/// This Continue button doesn't have a call in for `onNext` either, just to emphasize the issues of relying
/// on users to setup everything themselves before calling the widget.
ContinueButton(onPressed)
```

Reply:
The only way I was able to offset the logic to the `ContinueButton` is to pass 3 additional parameters, and making `onPressedCallback` optional:
```
  final Key elevatedButtonKey;
  final String? error;
  final bool enabled;
  final DynamicFuture? onPressedCallback;
  final bool isLoading;
  final SetIsLoading? setIsLoading;
  final SetError? setError;
  final VoidCallback? nextStep;

The `ElevatedButton`'s `onPressed` becomes:
```
              onPressed: enabled == true
                  ? () async {
                      if (setError != null) setError!(null);
                      if (setIsLoading != null) setIsLoading!(true);
                      try {
                        if (onPressedCallback != null) await onPressedCallback!();
                      } catch (error, stacktrace) {
                        if (setError != null) {
                          setError!(errorToString(error, stacktrace));
                        }
                      } finally {
                        if (setIsLoading != null) setIsLoading!(false);
                      }
                      if (error == null && nextStep != null) nextStep!();
                    }
                  : null,
                  ```
As you can see, the logic in `onPressed` becomes much more complicated. Because I have to accommodate for all the steps that are using the `ContinueButton`. Some steps do not have a `onPressedCallback`, and they simply executes `nextStep`. If they simply call `nextStep`, it is unnecessary for them to have `loading` or `error` (since `nextStep` does not load, and it can't throw any error), thus, there is no need to have `setError` or `setLoading` for these steps. That's why I have to do all these null checks.

The main problem is `isLoading`, `error` of `ContinueButton` must come from the parents, so does `setError`, and `nextStep`. Because parents needs to consume these states for other widgets. If the parent doesn't need these, then these states could simply be declared in `ContinueButton`, and there won't be any problem offloading all the logic to `ContinueButton`. 

It would be nicer to have the logic in `ContinueButton`, but the latter is used in many steps, and it should accommodate for all different behaviours. It doesn't appear to be beneficial to offload the logic to `ContinueButton`, unless you have a better approach. 

What do you think if we leave the logic in the parent as it is, and when all the continue buttons are made, I can revisit to see if there is an efficient way to offload all the logic used by each step to `ContinueButton`. It will be simpler to do when the logic for each step is made.